src/test/java/htsjdk/samtools/cram/structure/SliceTests.java

```diff
@@ -72,238 +75,194 @@ public void testFailsMD5Check() throws IOException {
 
-        records.add(single);
+    private CramCompressionRecord createRecordHalfUnmapped(final int index,
+                                                           final int sequenceId) {
+        final CramCompressionRecord record = createRecord(index, sequenceId);
+        if (index % 2 == 0) {
+            record.setSegmentUnmapped(true);
+        } else {
+            record.setSegmentUnmapped(false);
+        }
+        return record;
+    }
 
-        final CramCompressionRecord unmapped = new CramCompressionRecord();
-        unmapped.readBases = "AAA".getBytes();
-        unmapped.qualityScores = "!!!".getBytes();
-        unmapped.readLength = 3;
-        unmapped.readName = "unmapped";
-        unmapped.sequenceId = SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX;
-        unmapped.alignmentStart = SAMRecord.NO_ALIGNMENT_START;
-        unmapped.setLastSegment(true);
-        unmapped.readFeatures = Collections.emptyList();
-        unmapped.setSegmentUnmapped(true);
+    private List<CramCompressionRecord> getSingleRefRecords() {
+        final List<CramCompressionRecord> records = new ArrayList<>();
+        for (int index = 0; index < TEST_RECORD_COUNT; index++) {
+            records.add(createRecordHalfUnmapped(index, 0));
+        }
+        return records;
+    }
 
-        records.add(unmapped);
+    private List<CramCompressionRecord> getMultiRefRecords() {
+        final List<CramCompressionRecord> records = new ArrayList<>();
+        for (int index = 0; index < TEST_RECORD_COUNT; index++) {
+            records.add(createRecordHalfUnmapped(index, index));
+        }
+        return records;
+    }
 
-        final CompressionHeader header = new CompressionHeaderFactory().build(records, null, true);
 
-        final Slice slice = Slice.buildSlice(records, header);
+    private List<CramCompressionRecord> getNoRefRecords() {
+        final List<CramCompressionRecord> records = new ArrayList<>();
+        for (int index = 0; index < TEST_RECORD_COUNT; index++) {
+            final CramCompressionRecord record = createRecord(index, SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX);
+            record.setSegmentUnmapped(true);
+            records.add(record);
+        }
+        return records;
+    }
 
-        Assert.assertEquals(slice.nofRecords, 2);
-        Assert.assertEquals(slice.sequenceId, Slice.MULTI_REFERENCE);
-        Assert.assertEquals(slice.alignmentStart, Slice.NO_ALIGNMENT_START);
-        Assert.assertEquals(slice.alignmentSpan, Slice.NO_ALIGNMENT_SPAN);
+    private List<CramCompressionRecord> getNoStartRecords() {
+        final List<CramCompressionRecord> records = new ArrayList<>();
+        for (int index = 0; index < TEST_RECORD_COUNT; index++) {
+            final CramCompressionRecord record = createRecord(index, index);
+            record.alignmentStart = SAMRecord.NO_ALIGNMENT_START;
+            record.setSegmentUnmapped(true);
+            records.add(record);
```

Comment:
The concern I was expressing in the last round is that these methods create a mix of records, some of which are in "undefined" territory (which I referred to as "half-x"). This method creates some records (half of them, but thats not what I meant by "half") that have the unmapped flag set; but have a valid reference index; and not a valid start. Although we want to be able to handle those, they're outside the normal range of records we expect to see. I'd like to find a way to make that explicit (or preferably, separate such data from the normal cases), or at least make it clear that its intentional. Let me know what you think about this issue or if it doesnt make sense.

Reply:
Oh I see now.  Yes, it is intentional - I want to cover all of these half-cases.  Would adding comments like this work for you?

> These records have valid startAlignments but don't have valid references.  This is not something we expect to see in reads, but we want to treat them as unplaced if we do.  These tests cover that rare case.

Reply:
That will work for the short term, sure.